File: skills/SKILL-11-consultation/ios-swiftui/patterns.swift
import SwiftUI

// MARK: - 1. ZONE DE TOUCHE [AA] — Critère 11.5
// Minimum 44×44 points (recommandation Apple HIG)
// ❌ Mauvais: icône 20×20pt sans zone de touche élargie
// ✅ Bon: .frame(minWidth: 44, minHeight: 44) avec .contentShape(Rectangle())

struct TouchTargetPattern: View {
    var body: some View {
        HStack(spacing: 20) {
            // Inaccessible : Zone de touche trop petite
            Button(action: { print("Action") }) {
                Image(systemName: "star")
                    .resizable()
                    .frame(width: 20, height: 20)
            }
            .accessibilityLabel("Mettre en favori (Trop petit)")
            
            // Accessible : Zone de touche étendue sans altérer le visuel
            Button(action: { print("Action") }) {
                Image(systemName: "star")
                    .resizable()
                    .frame(width: 20, height: 20)
            }
            .frame(minWidth: 44, minHeight: 44)
            .contentShape(Rectangle()) // Rend la zone de padding cliquable
            .accessibilityLabel("Mettre en favori")
        }
    }
}

// MARK: - 2. TIMEOUT [A] — Critère 11.1
// Timer de session: avertir avant expiration avec option de prolonger

struct SessionTimeoutPattern: View {
    @State private var isTimeoutWarningVisible = false
    @State private var timeRemaining = 120
    
    var body: some View {
        VStack {
            Text("Espace sécurisé")
            Button("Simuler expiration de session") {
                isTimeoutWarningVisible = true
                UIAccessibility.post(notification: .announcement, argument: "Votre session expire dans 2 minutes. Voulez-vous la prolonger ?")
            }
        }
        .alert(isPresented: $isTimeoutWarningVisible) {
            Alert(
                title: Text("Expiration de session imminente"),
                message: Text("Pour des raisons de sécurité, votre session expirera dans \(timeRemaining) secondes."),
                primaryButton: .default(Text("Prolonger la session")) {
                    resetSession()
                },
                secondaryButton: .cancel(Text("Me déconnecter")) {
                    logout()
                }
            )
        }
    }
    
    private func resetSession() {
        timeRemaining = 300 // Reset timer
    }
    
    private func logout() {}
}

// MARK: - 3. ERREURS RÉCUPÉRABLES [AA] — Critère 11.6
// Confirmation avant suppression irréversible via .confirmationDialog

struct RecoverableErrorPattern: View {
    @State private var isDeleteConfirmationPresented = false
    
    var body: some View {
        Button(role: .destructive) {
            isDeleteConfirmationPresented = true
        } label: {
            Text("Supprimer le compte")
                .frame(minWidth: 44, minHeight: 44) // Zone de touche
        }
        .confirmationDialog(
            "Êtes-vous sûr de vouloir supprimer définitivement votre compte ? Cette action est irréversible.",
            isPresented: $isDeleteConfirmationPresented,
            titleVisibility: .visible
        ) {
            Button("Supprimer", role: .destructive) {
                deleteAccount()
            }
            Button("Annuler", role: .cancel) {
                // Annulation, pas d'action
            }
        }
    }
    
    private func deleteAccount() {
        // Logique de suppression
    }
}

// MARK: - 4. ERREURS SAISIE RÉCUPÉRABLES [AA]
// Ne pas vider les champs valides après une erreur

struct InputRetentionPattern: View {
    @State private var username: String = ""
    @State private var email: String = ""
    @State private var errorMessage: String? = nil
    
    var body: some View {
        Form {
            Section(header: Text("Informations de profil")) {
                TextField("Nom d'utilisateur", text: $username)
                    .textContentType(.username)
                    .autocapitalization(.none)
                
                TextField("Email", text: $email)
                    .keyboardType(.emailAddress)
                    .textContentType(.emailAddress)
                    .autocapitalization(.none)
                
                if let errorMessage = errorMessage {
                    Text(errorMessage)
                        .foregroundColor(.red)
                        .accessibilityLabel("Erreur : \(errorMessage)")
                }
                
                Button("Sauvegarder") {
                    validateForm()
                }
                .frame(minHeight: 44) // Accessibilité zone de touche
            }
        }
    }
    
    private func validateForm() {
        if !email.contains("@") {
            // L'erreur est affichée, mais `username` et le début d'`email` ne sont PAS effacés
            errorMessage = "Veuillez entrer une adresse email valide."
            UIAccessibility.post(notification: .announcement, argument: errorMessage)
        } else {
            errorMessage = nil
        }
    }
}

File: skills/SKILL-11-consultation/android-compose/patterns.kt
package com.fram.a11y.skill11

import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.semantics.*
import androidx.compose.ui.unit.dp

// MARK: - 1. ZONE DE TOUCHE [AA] — Critère 11.5
// Minimum 48×48dp (Material Design recommandation)
// .minimumInteractiveComponentSize() ou .size(48.dp)

@Composable
fun TouchTargetPattern() {
    Row(horizontalArrangement = Arrangement.spacedBy(16.dp)) {
        // Inaccessible : trop petit, difficile à toucher (ex: 24.dp sans padding cliquable)
        IconButton(
            onClick = { /* Action */ },
            modifier = Modifier.size(24.dp) // Mauvais
        ) {
            Icon(painterResource(android.R.drawable.ic_menu_edit), contentDescription = "Éditer (Inaccessible)")
        }

        // Accessible : zone de touche minimale assurée (IconButton utilise minimumInteractiveComponentSize par défaut,
        // mais pour des custom views, il faut l'imposer)
        IconButton(
            onClick = { /* Action */ },
            modifier = Modifier.minimumInteractiveComponentSize() // Assure 48x48dp au minimum
        ) {
            Icon(painterResource(android.R.drawable.ic_menu_edit), contentDescription = "Éditer")
        }
    }
}

// MARK: - 2. TIMEOUT [A] — Critère 11.1
// AlertDialog avant expiration de session

@Composable
fun SessionTimeoutPattern() {
    var showTimeoutDialog by remember { mutableStateOf(false) }

    Column {
        Button(
            onClick = { showTimeoutDialog = true },
            modifier = Modifier.minimumInteractiveComponentSize()
        ) {
            Text("Simuler l'expiration")
        }
    }

    if (showTimeoutDialog) {
        AlertDialog(
            onDismissRequest = { /* Empêche la fermeture sans action explicite si sécurité critique */ },
            title = { 
                Text(
                    text = "Expiration de session imminente",
                    modifier = Modifier.semantics { liveRegion = LiveRegionMode.Assertive }
                ) 
            },
            text = { Text("Votre session expirera dans 2 minutes. Voulez-vous la prolonger ?") },
            confirmButton = {
                TextButton(
                    onClick = { 
                        showTimeoutDialog = false
                        // reset session logic
                    },
                    modifier = Modifier.minimumInteractiveComponentSize()
                ) {
                    Text("Prolonger")
                }
            },
            dismissButton = {
                TextButton(
                    onClick = { 
                        showTimeoutDialog = false
                        // logout logic
                    },
                    modifier = Modifier.minimumInteractiveComponentSize()
                ) {
                    Text("Me déconnecter")
                }
            }
        )
    }
}

// MARK: - 3. CONFIRMATION AVANT ACTION IRREVÉRSIBLE [AA]
// AlertDialog pour supprimer/annuler

@Composable
fun RecoverableErrorConfirmationPattern() {
    var showDeleteConfirmation by remember { mutableStateOf(false) }

    Button(
        onClick = { showDeleteConfirmation = true },
        colors = ButtonDefaults.buttonColors(containerColor = MaterialTheme.colorScheme.error),
        modifier = Modifier.minimumInteractiveComponentSize()
    ) {
        Text("Supprimer le compte")
    }

    if (showDeleteConfirmation) {
        AlertDialog(
            onDismissRequest = { showDeleteConfirmation = false },
            title = { Text("Supprimer définitivement ?") },
            text = { Text("Cette action est irréversible. Toutes vos données seront perdues.") },
            confirmButton = {
                Button(
                    onClick = { 
                        showDeleteConfirmation = false
                        // Delete logic
                    },
                    colors = ButtonDefaults.buttonColors(containerColor = MaterialTheme.colorScheme.error),
                    modifier = Modifier.minimumInteractiveComponentSize()
                ) {
                    Text("Oui, supprimer")
                }
            },
            dismissButton = {
                TextButton(
                    onClick = { showDeleteConfirmation = false },
                    modifier = Modifier.minimumInteractiveComponentSize()
                ) {
                    Text("Annuler")
                }
            }
        )
    }
}

// MARK: - 4. DONNÉES CONSERVÉES APRÈS ERREUR [AA]
// Ne pas réinitialiser les champs de formulaire

@Composable
fun InputRetentionPattern() {
    var email by remember { mutableStateOf("") }
    var isError by remember { mutableStateOf(false) }

    Column(modifier = Modifier.padding(16.dp)) {
        OutlinedTextField(
            value = email,
            onValueChange = { 
                email = it
                isError = false // Réinitialiser l'erreur quand l'utilisateur corrige
            },
            label = { Text("Adresse e-mail") },
            isError = isError,
            supportingText = {
                if (isError) {
                    Text(
                        text = "Veuillez entrer une adresse e-mail valide.",
                        color = MaterialTheme.colorScheme.error,
                        modifier = Modifier.semantics { liveRegion = LiveRegionMode.Polite }
                    )
                }
            },
            modifier = Modifier.fillMaxWidth()
        )
        
        Spacer(modifier = Modifier.height(16.dp))
        
        Button(
            onClick = {
                if (!email.contains("@")) {
                    isError = true
                    // La saisie utilisateur (email) reste présente, on ne la vide pas !
                }
            },
            modifier = Modifier.minimumInteractiveComponentSize()
        ) {
            Text("Valider")
        }
    }
}
